### PR TITLE
Zones overhaul

### PR DESCRIPTION
**Pull request by joshuavictorchen**
## Description

Overhaul tooling in the `reactor.zones` module and remove application-specific zoning logic, which may be implemented as desired in downstream `Interface`s.

This satisfies the main intent of https://github.com/terrapower/armi/discussions/811. Enhancements may be added in the future.

Related issues:
https://github.com/terrapower/armi/issues/563
https://github.com/terrapower/armi/issues/800

---

## Checklist

- [x] This PR has only one purpose or idea.
- [x] Tests have been added/updated to verify that the new/changed code works.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/release/index.html) (location `doc/release/0.X.rst`) are up-to-date with any bug fixes or new features.
- [x] The documentation is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `setup.py`.

### CODE REVIEW

**Comment by john-science** [comment]
TODO:

* We need to bump the ARMI version number.
* We need to mention this in [the release docs](https://github.com/terrapower/armi/blob/main/doc/release/0.2.rst)
* In a future PR??? we should add some docs on how to define Zones.

**Comment by john-science** [comment]
@sombrereau @drewejohnson 

We are moving the super custom zones logic out of ARMI, so ARMI only provides a _way_ to define Zones.

I assume this is better for both of your reactor projects, but I just want to ask you.

File: armi/physics/fuelCycle/fuelHandlers.py [hunk 1]
```diff
@@ -724,6 +724,7 @@ def getParamWithBlockLevelMax(a, paramName):
                         # this assembly is in the excluded location list. skip it.
                         continue
 
+                # TODO: I'm pretty sure this is broken.
```
Comment (john-science) [inline armi/physics/fuelCycle/fuelHandlers.py:727]: We added this `TODO`.   What's up with this?

Comment (joshuavictorchen) [reply armi/physics/fuelCycle/fuelHandlers.py:727]: I don't remember inspecting this module, but the use of the `zones` tooling itself looks fine to me.

Comment (john-science) [reply armi/physics/fuelCycle/fuelHandlers.py]: Make sense, I probably added the TODO.

I removed it. Thanks!

**Review by john-science — APPROVED** [review]
@joshuavictorchen  @opotowsky   I think this is ready!  Do you think the downstream repos are ready?

**Comment by joshuavictorchen** [comment]
> @joshuavictorchen @opotowsky I think this is ready! Do you think the downstream repos are ready?

Yep! Zoning board meeting has been scheduled for the downstream application/repos :)

**Pull request merged**